File: MR.py
######################################Problem 1##############################
#Multinomial Regression
import pandas as pd
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import LabelEncoder

mdata = pd.read_csv("C:/Users/hp/Desktop/Multinomial R assi/mdata.csv")
mdata.head(10)

#EDA
mdata.describe()
mdata.prog.value_counts()
mdata.columns

#dropping unwanted columns
mdata.drop(mdata.iloc[:,[0,1]] ,axis = 1, inplace = True)

#label encoding 
#converting into binary
lb = LabelEncoder()
mdata["female"] = lb.fit_transform(mdata["female"])
mdata["ses"] = lb.fit_transform(mdata["ses"])
mdata["schtyp"] = lb.fit_transform(mdata["schtyp"])
mdata["honors"] = lb.fit_transform(mdata["honors"])

# Boxplot of independent variable distribution for each category of choice 
sns.boxplot(x = "prog", y = "female", data = mdata)
sns.boxplot(x = "prog", y = "ses", data = mdata)
sns.boxplot(x = "prog", y = "schtyp", data = mdata)
sns.boxplot(x = "prog", y = "read", data = mdata)
sns.boxplot(x = "prog", y = "write", data = mdata)
sns.boxplot(x = "prog", y = "math", data = mdata)
sns.boxplot(x = "prog", y = "science", data = mdata)
sns.boxplot(x = "prog", y = "honors", data = mdata)

# Scatter plot for each categorical choice of car
sns.stripplot(x = "prog", y = "female", jitter = True, data = mdata)
sns.stripplot(x = "prog", y = "ses", jitter = True, data = mdata)
sns.stripplot(x = "prog", y = "schtyp", jitter = True, data = mdata)
sns.stripplot(x = "prog", y = "read", jitter = True, data = mdata)
sns.stripplot(x = "prog", y = "write", jitter = True, data = mdata)
sns.stripplot(x = "prog", y = "math", jitter = True, data = mdata)
sns.stripplot(x = "prog", y = "science", jitter = True, data = mdata)
sns.stripplot(x = "prog", y = "honors", jitter = True, data = mdata)

# Scatter plot between each possible pair of independent variable and also histogram for each independent variable 
sns.pairplot(mdata) # Normal
sns.pairplot(mdata, hue = "prog") # With showing the category of each car choice in the scatter plot

# Correlation values between each independent features
mdata.corr()

train, test = train_test_split(mdata, test_size = 0.2)

# ‘multinomial’ option is supported only by the ‘lbfgs’ and ‘newton-cg’ solvers
model = LogisticRegression(multi_class = "multinomial", solver = "newton-cg").fit(train.iloc[:, [0,1,2,4,5,6,7,8]], train.iloc[:, 3])

test_predict = model.predict(test.iloc[:, [0,1,2,4,5,6,7,8]]) # Test predictions

# Test accuracy 
accuracy_score(test.iloc[:,3], test_predict)

train_predict = model.predict(train.iloc[:, [0,1,2,4,5,6,7,8]]) # Train predictions 
# Train accuracy 
accuracy_score(train.iloc[:,3], train_predict) 

#############################################Problem 2####################################
#Multinomial Regression
import pandas as pd
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import LabelEncoder

loan = pd.read_csv("C:/Users/hp/Desktop/Multinomial R assi/loan.csv")

#subsetting thr dataset
loan = loan.iloc[:,:17]

#Dropping unwanted columns
loan.drop(loan.iloc[:, [0,1,10,11,15]] , axis = 1 , inplace = True)
loan.head(10)

#EDA
loan.describe()
loan.loan_status.value_counts()
loan.columns

#label encoding 
#converting into binary
lb = LabelEncoder()
loan["term"] = lb.fit_transform(loan["term"])
loan["grade"] = lb.fit_transform(loan["grade"])
loan["sub_grade"] = lb.fit_transform(loan["sub_grade"])
loan["home_ownership"] = lb.fit_transform(loan["home_ownership"])
loan["verification_status"] = lb.fit_transform(loan["verification_status"])

#converting percentage value to float for 'int_rate' column
loan['int_rate'] = loan['int_rate'].str.rstrip('%').astype(float)

# Boxplot of independent variable distribution for each category of choice 
sns.boxplot(x = "loan_status", y = "loan_amnt", data = loan)
sns.boxplot(x = "loan_status", y = "funded_amnt", data = loan)
sns.boxplot(x = "loan_status", y = "funded_amnt_inv", data = loan)
sns.boxplot(x = "loan_status", y = "term", data = loan)
sns.boxplot(x = "loan_status", y = "int_rate", data = loan)
sns.boxplot(x = "loan_status", y = "installment", data = loan)
sns.boxplot(x = "loan_status", y = "grade", data = loan)
sns.boxplot(x = "loan_status", y = "sub_grade", data = loan)
sns.boxplot(x = "loan_status", y = "home_ownership", data = loan)
sns.boxplot(x = "loan_status", y = "annual_inc", data = loan)
sns.boxplot(x = "loan_status", y = "verification_status", data = loan)

# Scatter plot for each categorical choice of car
sns.stripplot(x = "loan_status", y = "loan_amnt", jitter = True, data = loan)
sns.stripplot(x = "loan_status", y = "funded_amnt", jitter = True, data = loan)
sns.stripplot(x = "loan_status", y = "funded_amnt_inv", jitter = True, data = loan)
sns.stripplot(x = "loan_status", y = "term", jitter = True, data = loan)
sns.stripplot(x = "loan_status", y = "int_rate", jitter = True, data = loan)
sns.stripplot(x = "loan_status", y = "installment", jitter = True, data = loan)
sns.stripplot(x = "loan_status", y = "grade", jitter = True, data = loan)
sns.stripplot(x = "loan_status", y = "sub_grade", jitter = True, data = loan)
sns.stripplot(x = "loan_status", y = "home_ownership", jitter = True, data = loan)
sns.stripplot(x = "loan_status", y = "annual_inc", jitter = True, data = loan)
sns.stripplot(x = "loan_status", y = "verification_status", jitter = True, data = loan)

# Scatter plot between each possible pair of independent variable and also histogram for each independent variable 
sns.pairplot(loan) # Normal
sns.pairplot(loan, hue = "loan_status") # With showing the category of each car choice in the scatter plot

# Correlation values between each independent features
loan.corr()

train, test = train_test_split(loan, test_size = 0.2)

# ‘multinomial’ option is supported only by the ‘lbfgs’ and ‘newton-cg’ solvers
model = LogisticRegression(multi_class = "multinomial", solver = "newton-cg").fit(train.iloc[:, :11], train.iloc[:, 11])

test_predict = model.predict(test.iloc[:, :11]) # Test predictions

# Test accuracy 
accuracy_score(test.iloc[:,11], test_predict)

train_predict = model.predict(train.iloc[:, :11]) # Train predictions 
# Train accuracy 
accuracy_score(train.iloc[:,11], train_predict)

#########################################END##########################################

File: MR.r
######################################Problem 1#########################
# Multinomial Logit Model
install.packages("mlogit")
require('mlogit')
install.packages("nnet")
require('nnet')
library(readr)

#loading dataset
mdata <- read_csv("C:\\Users\\hp\\Desktop\\Multinomial R assi\\mdata.csv")

#dropping unwanted columns
mdata <- mdata[,-c(1,2)]
head(mdata)
tail(mdata)
summary(mdata)

#getting count of response variable
table(mdata$prog) 

# Data Partitioning
n <-  nrow(mdata)
n1 <-  n * 0.85
n2 <-  n - n1
train_index <- sample(1:n, n1)
train <- mdata[train_index, ]
test <-  mdata[-train_index, ]

commute <- multinom(prog ~ female + ses + schtyp + read + write + math + science + honors, data = train)
summary(commute)

#Significance of Regression Coefficients
z <- summary(commute)$coefficients / summary(commute)$standard.errors

p_value <- (1 - pnorm(abs(z), 0, 1)) * 2

summary(commute)$coefficients
p_value

# odds ratio 
exp(coef(commute))

# check for fitted values on training data
prob <- fitted(commute)

# Predicted on test data
pred_test <- predict(commute, newdata =  test, type = "probs") # type="probs" is to calculate probabilities
pred_test

# Find the accuracy of the model
class(pred_test)
pred_test <- data.frame(pred_test)
View(pred_test)
pred_test["prediction"] <- NULL

# Custom function that returns the predicted value based on probability
get_names <- function(i){
  return (names(which.max(i)))
}

predtest_name <- apply(pred_test, 1, get_names)
pred_test$prediction <- predtest_name
View(pred_test)

# Confusion matrix
table(predtest_name, test$prog)

# confusion matrix visualization
barplot(table(predtest_name, test$prog), beside = T, col =c("red", "lightgreen", "blue", "orange"), main = "Predicted(X-axis) - Legends(Actual)", ylab ="count")

# Accuracy on test data
mean(predtest_name == test$prog)

# Training Data
pred_train <- predict(commute, newdata =  train, type="probs") # type="probs" is to calculate probabilities
pred_train

# Find the accuracy of the model
class(pred_train)
pred_train <- data.frame(pred_train)
View(pred_train)
pred_train["prediction"] <- NULL

predtrain_name <- apply(pred_train, 1, get_names)
pred_train$prediction <- predtrain_name
View(pred_train)

# Confusion matrix
table(predtrain_name, train$prog)

# confusion matrix visualization
barplot(table(predtrain_name, train$prog), beside = T, col =c("red", "lightgreen", "blue", "orange"), main = "Predicted(X-axis) - Legends(Actual)", ylab ="count")

# Accuracy 
mean(predtrain_name == train$prog)

#######################################Problem 2###########################################
# Multinomial Logit Model
#install.packages("mlogit")
require('mlogit')
#install.packages("nnet")
require('nnet')
library(readr)

#loading dataset
loan <- read_csv("C:\\Users\\hp\\Desktop\\Multinomial R assi\\loan.csv")

#subsetting the csv file
loan <- loan[,1:17]

#dropping unwanted columns
loan <- loan[,-c(1,2,11,12,16)]

#converting 'init_rate' to float (removing %)
loan$int_rate <- sapply(loan$int_rate, gsub , pattern = '%' , replacement = '')

head(loan)
tail(loan)
names(loan)

#getting count of response variable
table(loan$loan_status)

# Data Partitioning
loan <- loan[1:1000,]
n <-  nrow(loan)
n1 <-  n * 0.85
n2 <-  n - n1
train_index <- sample(1:n, n1)
train <- loan[train_index, ]
test <-  loan[-train_index, ]

commute <- multinom(loan_status ~ ., data = train)
summary(commute)
#loan_amnt + funded_amnt + funded_amnt_inv + term + int_rate + installment + grade + sub_grade + home_ownership + annual_inc + verification_status
#Significance of Regression Coefficients
z <- summary(commute)$coefficients / summary(commute)$standard.errors

p_value <- (1 - pnorm(abs(z), 0, 1)) * 2

summary(commute)$coefficients
p_value

# odds ratio 
exp(coef(commute))

# check for fitted values on training data
prob <- fitted(commute)

# Predicted on test data
pred_test <- predict(commute, newdata =  test, type = "probs") # type="probs" is to calculate probabilities
pred_test

# Find the accuracy of the model
class(pred_test)
pred_test <- data.frame(pred_test)
View(pred_test)
pred_test["prediction"] <- NULL

# Custom function that returns the predicted value based on probability
get_names <- function(i){
  return (names(which.max(i)))
}

predtest_name <- apply(pred_test, 1, get_names)
pred_test$prediction <- predtest_name
View(pred_test)

# Confusion matrix
table(predtest_name, test$loan_status)

# confusion matrix visualization
barplot(table(predtest_name, test$loan_status), beside = T, col =c("red", "lightgreen", "blue", "orange"), main = "Predicted(X-axis) - Legends(Actual)", ylab ="count")

# Accuracy on test data
mean(predtest_name == test$loan_status)

# Training Data
pred_train <- predict(commute, newdata =  train, type="probs") # type="probs" is to calculate probabilities
pred_train

# Find the accuracy of the model
class(pred_train)
pred_train <- data.frame(pred_train)
View(pred_train)
pred_train["prediction"] <- NULL

predtrain_name <- apply(pred_train, 1, get_names)
pred_train$prediction <- predtrain_name
View(pred_train)

# Confusion matrix
table(predtrain_name, train$loan_status)

# confusion matrix visualization
barplot(table(predtrain_name, train$loan_status), beside = T, col =c("red", "lightgreen", "blue", "orange"), main = "Predicted(X-axis) - Legends(Actual)", ylab ="count")

# Accuracy 
mean(predtrain_name == train$loan_status)

######################################END###########################################